### PR TITLE
Compo metrics

### PR DESCRIPTION
**Pull request by TheSpaceSheep**
Combining the features of the 4 previous pull requests in one branch.
- layernorm
- testing on 1, 2, ..., 10 revealed
- finer baseline for masked inputs
- compositionality metrics for masked input

Add new mask distributions
 - heat parameter to go from deterministic to uniform
 - binomial distributions, that evolves with the accuracy

### CODE REVIEW

File: egg/core/baselines.py [hunk 1]
```diff
@@ -87,3 +87,34 @@ def predict(self, loss: torch.Tensor) -> torch.Tensor:
...
+            return loss
+
+        mask = aux_input["mask"]
+        
+        bsz = loss.size(0)
+
+        loss_detached = loss.detach()
+        mean = torch.zeros_like(loss_detached)
+        for i in range(1, mask.shape[1]+1):
+            n_masks = mask.sum(1)
```
Comment (TheSpaceSheep) [inline egg/core/baselines.py]: This can be done before the for loop

File: egg/zoo/compo_vs_generalization/intervention.py [hunk 2]
```diff
@@ -117,47 +122,86 @@ def cosine_dist(_list):
...
+        corr = spearmanr(distance_messages, distance_inputs).correlation
+        top_sims[batch_id] = corr
+
+    return top_sims.mean().item()
 
 
 class Metrics(core.Callback):
-    def __init__(self, dataset, device, n_attributes, n_values, vocab_size, freq=1):
-        self.dataset = dataset
+    def __init__(self, dataset_small, device, n_attributes, n_values, vocab_size, freq=1, rng=None):
```
Comment (TheSpaceSheep) [inline egg/zoo/compo_vs_generalization/intervention.py]: revert the name of argument to "dataset" and not "dataset_small"

File: egg/zoo/compo_vs_generalization/intervention.py [hunk 3]
```diff
@@ -117,47 +122,86 @@ def cosine_dist(_list):
...
 
     def dump_stats(self):
         game = self.trainer.game
         game.eval()
 
-        positional_disent = information_gap_position(
-            self.n_attributes, self.n_values, self.dataset, game.sender, self.device
-        )
-        bos_disent = information_gap_vocab(
+        #positional_disent = information_gap_position(
```
Comment (TheSpaceSheep) [inline egg/zoo/compo_vs_generalization/intervention.py]: uncomment this

**Review by TheSpaceSheep — COMMENTED** [review]
(self review) found some minor mistakes to correct
